### ISSUE TITLE
library files can't go in `files/` segment -- broken with no_lazy_load set to false

### ISSUE THREAD

**Issue by lamont-granquist** [chef-cookbooks/compat_resource#10]
This is at least the first issue which breaks Chef 11.x users, but is also going to be an issue for any customer which is using Chef 12.x and has set no_lazy_load back to false again (which is still a supported mode since the default has performance implications).

The lib files need to get moved, probably back into the `libraries/` segment.  Since we've loaded libraries recursively since 12.0.0.rc.0 this will work for Chef 12.x users:

https://github.com/chef/chef/commit/18b555fb4dbf848ebfd274a5ceeda1b6a2427efd

It won't fix 11.x users, but I kinda doubt that they're fixable.

**Comment by jkeiser** [comment]
I'm on board with moving to libraries. I hadn't thought that library load was recursive; I agree it'll probably work just fine.

**Comment by jkeiser** [comment]
local mode has some issues with recursive libraries, it seems; gotta fix that before we can move.

**Comment by jkeiser** [comment]
There is also the issue that autoloading the entire library can be the wrong thing to do--if the actual gem has already been loaded.

**Comment by lamont-granquist** [comment]
i'm still :-1: on using this as a gem and feel that cheffish and chef_provisioning need to be moved into core and/or tied to core exactly like chef-config is.

https://github.com/chef/cheffish/pull/100

the feature of running this as a gem is much more pain that its worth.

**Comment by lamont-granquist** [comment]
so the issue with chef-zero is probably a show-stopper for using recursive libraries files.

we have to be able to support omnibus-chef for 12.0.3 or whatever along with the chef-zero that shipped in that package, since chef-zero only recently got fixed to do recursive library loading we can't rely on that feature to be present (even if we monkeypatch it we have a chicken-and-egg issue because we're going to be monkeypatching a method which has already run to do the synchronization of the compat_resources cookbook.

**Comment by lamont-granquist** [comment]
This is probably not fixable, so this cookbook will most likely just remain busted if no_lazy_load is false.  No actual consumers have reported this.  If it really does come up and needs fixing the way to fix it is most likely to do something when the library file loads up to trigger the loading of all the files in the manifest.
